src/core/lib/security/credentials/tls/grpc_tls_credentials_options.cc

```diff
@@ -0,0 +1,203 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+
+/** -- gRPC TLS key materials config API implementation. -- **/
+void grpc_tls_key_materials_config::set_key_materials(
+    const grpc_ssl_pem_key_cert_pair* key_cert_pairs, const char* root_certs,
+    size_t num) {
+  if (key_cert_pairs == nullptr || num == 0) {
+    gpr_log(GPR_ERROR, "Invalid arguments to set_key_materials()");
+    return;
+  }
+  pem_key_cert_pairs_ = static_cast<grpc_ssl_pem_key_cert_pair*>(
```

Comment:
You may want to free pem_key_cert_pairs_ and pem_root_certs_ first. Otherwise, it will have memory leak if set_key_materials is called twice.

Reply:
Good points. The code is revised.